sympy/functions/elementary/hyperbolic.py

```diff
@@ -1180,16 +1180,24 @@ def taylor_term(n, x, *previous_terms):
                 return S.NegativeOne**k * R / F * x**n / n
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
-        from sympy.series.order import Order
-        arg = self.args[0].as_leading_term(x)
-
-        if x in arg.free_symbols and Order(1, x).contains(arg):
-            return arg
-        else:
-            return self.func(arg)
+        arg = self.args[0]
+        x0 = arg.subs(x, 0).cancel()
+        if x0.is_zero:
+            return arg.as_leading_term(x)
+        # Handling branch points (-I*oo, -I) U (I, I*oo)
+        if x0 in (-S.ImaginaryUnit, S.ImaginaryUnit, S.ComplexInfinity):
+            return self.rewrite(log)._eval_as_leading_term(x, logx=logx, cdir=cdir)
+        if cdir != 0:
+            cdir = arg.dir(x, cdir)
+        if re(cdir) > 0 and re(x0).is_zero and im(x0) < S.NegativeOne:
```

Comment:
Conditions like `im(x0) < S.NegativeOne` do not necessarity return a boolean value.
```
In [6]: r = symbols('r', real=True)                                            

In [7]: e = asinh(I*r + x)                                                     

In [8]: e._eval_as_leading_term(x, None, 1)                                    
---------------------------------------------------------------------------
...
TypeError: cannot determine truth value of Relational
```

Reply:
It will be better to use a condition that will not raise an error. There are several possibilities, e.g., `(im(x0) -1).is_negative`, or there could be something like `-I*pi*Heaviside(1 - im(x0), 0)` in the return value. 
It seems that some experimentation is necessary.

Reply:
Yeah , even I was thinking about `Heaviside` , will look into this and explore the options we have .

Reply:
If we plan to use Heaviside though so generic cases like the one above which you've pointed out involving r , we would still need to differentiate based on cdir. For eg
```
>>> x = Symbol('x')
>>> r = symbols('r', real=True)
>>> asinh(2*I + x)._eval_as_leading_term(x, cdir = 1)
-I*asin(2)
>>> asinh(-2*I + x)._eval_as_leading_term(x, cdir = 1)
-I*pi + I*asin(2)
>>> asinh(r*I + x)._eval_as_leading_term(x, cdir = 1)
-I*pi*Heaviside(1 - r, 0) - I*asin(r)
>>> expr = _
>>> expr.subs(r, 2)
-I*asin(2)
>>> expr.subs(r, -2)
-I*pi + I*asin(2)
>>>
>>>
>>> asinh(2*I + x)._eval_as_leading_term(x, cdir = -1)
-I*asin(2) + I*pi
>>> asinh(-2*I + x)._eval_as_leading_term(x, cdir = -1)
I*asin(2)
>>> asinh(r*I + x)._eval_as_leading_term(x, cdir = -1)
I*pi*Heaviside(r + 1, 0) - I*asin(r)
>>> expr = _
>>> expr.subs(r, 2)
-I*asin(2) + I*pi
>>> expr.subs(r, -2)
I*asin(2)
```
So it's not that we can't get the correct answers , but the same Heaviside function won't work for both `cdir`. So a diff like this could be the case
```
+        if re(cdir) > 0 and re(x0).is_zero and (im(x0) + 1).is_negative:
             return -self.func(x0) - I*pi
-        elif re(cdir) < 0 and re(x0).is_zero and im(x0) > S.One:
+        elif re(cdir) < 0 and re(x0).is_zero and (im(x0) - 1).is_positive:
             return -self.func(x0) + I*pi
+        if re(cdir) > 0:
+            return -self.func(x0) -I*pi*Heaviside(1 - im(x0), 0)
+        if re(cdir) < 0:
+            return -self.func(x0) + I*pi*Heaviside(1 + im(x0), 0)
```

Reply:
Heaviside is probably overkill here. How about using rewrite(log) as in the case of trigonometric functions (possibly only when `re(x0).is_zero`)?

BTW, this would be easier to review if the name of the function were added in a comment like

    def _eval_as_leading_term(self, x, logx=None, cdir=0):  # asinh

Reply:
This is what finally what it pans out to for `asinh` from my side ( after carefully reviewing all cases)
```
    def _eval_as_leading_term(self, x, logx=None, cdir=0):
        from sympy import Piecewise, Heaviside
        arg = self.args[0]
        x0 = arg.subs(x, 0).cancel()
        if x0.is_zero:
            return arg.as_leading_term(x)
        # Handling branch points (-I*oo, -I) U (I, I*oo)
        if x0 in (-S.ImaginaryUnit, S.ImaginaryUnit, S.ComplexInfinity):
            return self.rewrite(log)._eval_as_leading_term(x, logx=logx, cdir=cdir)
        if cdir != 0:
            cdir = arg.dir(x, cdir)
        if re(x0).is_zero:
            if re(cdir) > 0:
                return Piecewise((-self.func(x0) - I*pi*Heaviside(-1 - im(x0), 1), Eq(Heaviside(-1 - im(x0), 1), S.One)),
                                 (self.func(x0), True))
            if re(cdir) < 0:
                return Piecewise((-self.func(x0) + I*pi*Heaviside(-1 - im(x0), 1), Eq(Heaviside(im(x0) - 1, 1), S.One)),
                                 (self.func(x0), True))
        return self.func(x0)
```
I feel we could have maybe tried this , the general cases would surely need condition or `Heaviside` . But yeah this is surely not required , as these would get much more complicated for cases involving `acoth, asech, acsch` (which have cases like `(x0) < S.One and (x0) > S.Zero`) and then we would need two `Heaviside` base conditions to implement this check which we should be avoiding !

Reply:
> BTW, this would be easier to review if the name of the function were added in a comment like

Sure will do !

Yeah I had tried log rewrites in this comment above (https://github.com/sympy/sympy/pull/23715#discussion_r913782499) . Though it expresses a lot of cases , we would also miss out of some cases based on `cdir` is what I had deduced. Though I'll try out what you've suggested !

Reply:
Would the (last) rewrite line alone suffice when re(x0).is_zero?

Reply:
In that case we could just do something like the following for all the cases :) and everything would work fine 
```
    def _eval_as_leading_term(self, x, logx=None, cdir=0):
        return self.rewrite(log)._eval_as_leading_term(x, logx=logx, cdir=cdir)
```

But like this would make the function computationally expensive ! So I would say the general cases should like the ones you've pointed out should be done using `self.func(x0)` which is more than enough for the doing the numerical cases , yeah but if we have something where we deal with a variable like `r` which can cover all set of real numbers, yeah then definitly in that case we should use log rewrite as I've shown in the diff !

Reply:
>  this would make the function computationally expensive

That may be true. Hence I think that rewrite(log) could be used only in some cases like x0.is_imaginary (or re(x0) == 0).

Reply:
Okay, maybe we could try this .As I wrote above , we don't really need the log rewrite for all `x0` where `re(x0) == 0`, most of the cases would suffice using `self.func(x0)` . Just the case involving `r` might be needing a log rewrite , but  yeah we can try that . 
So should I do this for all inverse trig/hyperbolic functions ? Like each of these have a block like which is basically responsible for handling points inside the branch cuts (and not on them) like for `asinh`
```
        if cdir != 0:
            cdir = arg.dir(x, cdir)
        if re(cdir) > 0 and re(x0).is_zero and im(x0) < S.NegativeOne:
            return -self.func(x0) - I*pi
        elif re(cdir) < 0 and re(x0).is_zero and im(x0) > S.One:
            return -self.func(x0) + I*pi
```
We have such a block in all inverse functions . If we want to replace this by a log rewrite we could maybe try that all functions ! Our functions (eg asinh) would look like this after the change
```
    def _eval_as_leading_term(self, x, logx=None, cdir=0):
        arg = self.args[0]
        x0 = arg.subs(x, 0).cancel()
        if x0.is_zero:
            return arg.as_leading_term(x)
        if re(x0).is_zero or x0 is S.S.ComplexInfinity:
            return self.rewrite(log)._eval_as_leading_term(x, logx=logx, cdir=cdir)
        return self.func(x0)
```
Though I still say that all cases (specially where `im(x0) > 1` and `im(x0) < -1`) don't need a log rewrite as it's only at branch cuts where sympy find it difficult to deal with series/limits , otherwise `self.func(x0)` is good enough ! 
Am I missing something here what you're pointing out , how is a simple log rewrite beneficial for all cases where `re(x0) == 0` ? 
I am only able to see a usecase in the example above using `r` that too not fully sure about that as it is won't work for all cases .
I like this block better where we keep log rewrite as the last resort .
```
        if re(x0).is_zero:
            if re(cdir) > 0 and (im(x0) + 1).is_negative:
                return -self.func(x0) - I*pi
            elif re(cdir) < 0 and (im(x0) - 1).is_positive:
                return -self.func(x0) + I*pi
            else:
                return self.rewrite(log)._eval_as_leading_term(x, logx=logx, cdir=cdir)
```
I guess I maybe missing something here ?!

Reply:
What will happen on the branch cut when `arg.dir(x, cdir)` is purely imaginary as in `asinh(2*I + I*x - x**2)`?

Reply:
This seems to be a really nice catch , my patch and log rewrites both work for `asinh(2*I - x**2)` and `asinh(2*I + x - x**2)` but not for `asinh(2*I + I*x - x**2)`
```
>>> expr = asinh(2*I + I*x - x**2)
>>> expr.as_leading_term(x, cdir = 1)
 I*asin(2)
>>> _.n()
1.31695789692482 + 1.5707963267949*I
>>> expr.rewrite(log).as_leading_term(x, cdir = 1)
log(sqrt(3) + 2) + I*pi/2
>>> _.n()
1.31695789692482 + 1.5707963267949*I
```
The expected answer here is 
```
>>> expr.as_leading_term(x, cdir = 1)
I*pi - asin(2)
>>> _.n()
-1.31695789692482 + 1.5707963267949*I
>>> expr.subs(x, 0.00000000000000001)
-1.31695789692482 + 1.5707963267949*I
```
I'm somewhat surprised that both methods failed here , specially the log one. I started looking into it , as of now I haven't been able to locate what are we missing .
```
>>> expr.rewrite(log)
log(-x**2 + I*x + sqrt((-x**2 + I*x + 2*I)**2 + 1) + 2*I)
```